src/jetstream_push_subscription.rs

```diff
@@ -0,0 +1,544 @@
+// Copyright 2020-2021 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+use crossbeam_channel as channel;
+
+use crate::jetstream::JetStream;
+use crate::jetstream_types::*;
+use crate::message::Message;
+use crate::DEFAULT_FLUSH_TIMEOUT;
+
+#[derive(Debug)]
+pub(crate) struct Inner {
+    /// Subscription ID.
+    pub(crate) sid: AtomicU64,
+
+    /// MSG operations received from the server.
+    pub(crate) messages: channel::Receiver<Message>,
+
+    /// Name of the stream associated with the subscription.
+    pub(crate) stream: String,
+
+    /// Name of the consumer associated with the subscription.
+    pub(crate) consumer: String,
+
+    /// Ack policy used in while processing messages.
+    pub(crate) consumer_ack_policy: AckPolicy,
+
+    /// Indicates if we own the consumer and are responsible for deleting it or not.
+    pub(crate) consumer_ownership: ConsumerOwnership,
+
+    /// Client associated with subscription.
+    pub(crate) context: JetStream,
+}
+
+impl Drop for Inner {
+    fn drop(&mut self) {
+        self.context
+            .connection
+            .0
+            .client
+            .unsubscribe(self.sid.load(Ordering::Relaxed))
+            .ok();
+    }
+}
+
+/// A `PushSubscription` receives `Message`s published
+/// to specific NATS `Subject`s.
+#[derive(Clone, Debug)]
+pub struct PushSubscription(pub(crate) Arc<Inner>);
+
+impl PushSubscription {
+    /// Creates a subscription.
+    pub(crate) fn new(
+        sid: u64,
+        consumer_info: ConsumerInfo,
+        consumer_ownership: ConsumerOwnership,
+        messages: channel::Receiver<Message>,
+        context: JetStream,
+    ) -> PushSubscription {
+        PushSubscription(Arc::new(Inner {
+            sid: AtomicU64::new(sid),
+            stream: consumer_info.stream_name,
+            consumer: consumer_info.name,
+            consumer_ack_policy: consumer_info.config.ack_policy,
+            consumer_ownership,
+            messages,
+            context,
+        }))
+    }
+
+    /// Filters and responds to protocol messages returning false if a message
+    /// should be filtered out.
+    fn filter(&self, message: &Message) -> bool {
```

Comment:
The naming is confusing.
I had to  check comment to understand when filter returns true.

As this is very specific filter, wouldn't specific name fit better?
i.e: `is_proto_message` ?
It would make the below if on `next()` also more read-friendly.

Reply:
Open to suggestions.
It has side effects tho, so naming shouldn't be a `is_` predicate.

Reply:
I have not idea what this is supposed to do? Consumers can be configured to have a filtered subject, and they actually should have one in all client impl. But I have not idea what this fn would be doing?

Reply:
Filters and handles protocol messages, e.g flow control.

Reply:
ok, so this can process HBs and FCs. Got it. Might have a better name to avoid confusion.

A flow like..

```rust
if control_msg(&m) {
  process_control{}
} else {
  do_callback()
```
}